batavia/VirtualMachine.js

```diff
@@ -1716,16 +1720,43 @@ VirtualMachine.prototype.byte_POP_EXCEPT = function() {
     this.unwind_block(block)
 }
 
-// VirtualMachine.prototype.byte_SETUP_WITH = function(dest) {
-//         ctxmgr = this.pop()
-//         this.push(ctxmgr.__exit__)
-//         ctxmgr_obj = ctxmgr.__enter__()
-//         if PY2:
-//             this.push_block('with', dest)
-//         elif PY3:
-//             this.push_block('finally', dest)
-//         this.push(ctxmgr_obj)
-// }
+VirtualMachine.prototype.byte_SETUP_WITH = function(dest) {
+    var mgr = this.pop()
+    this.push(mgr.__exit__) // XXX: native vs user defined object
+    var res = callables.call_method(mgr, '__enter__', [])
+    this.push_block('finally', dest)
+    this.push(res)
+}
+
+VirtualMachine.prototype.byte_WITH_CLEANUP = function() {
+    var exc = this.top()
+    var exit_func
+    var val, tb
+    if (exc instanceof types.NoneType) {
+        exit_func = this.pop(1)
+    } else if (exc instanceof String) {
+        if (exc === 'return' || exc === 'continue') {
+            exit_func = this.pop(2)
+        } else {
+            exit_func = this.pop(1)
+        }
+    } else if (exc.$pyclass.prototype instanceof exceptions.BaseException.$pyclass) {
+        val = this.peek(1)
+        tb = this.peek(2)
+        exit_func = this.pop(6)
+        this.frame.stack.splice(-3, 0, builtins.None)  // XXX
+        var block = this.pop_block()
+        this.push_block(block.type, block.handler, block.level - 1)
+    } else {
+        throw new builtins.BataviaError.$pyclass('Confused WITH_CLEANUP')
+    }
```

Comment:
I fixed it. The `XXX` was only there to remind me to double check that this line works as intended. But now I am wondering if I should leave it as is, since it breaks "stack API" by using `this.frame.stack.slice` instead of combination of `this.pop` and `this.push`. The other way feels clunky in this case.